### PR TITLE
i.sentinel_1 addon for Sentinel-1 SAR data processing

### PR DESCRIPTION
**Pull request by griembauer**
`i.sentinel_1` is a GRASS GIS addon toolset that searches, downloads, imports and preprocesses SAR data from the <a href="https://asf.alaska.edu/">Alaska Satellite Facility</a> using ESA's <a href="https://step.esa.int/main/download/snap-download/">SNAP</a> software. If a region is covered by multiple Sentinel-1 GRD tiles, a mosaic can be created.

The toolset consists of:

- i.sentinel_1.download_asf
- i.sentinel_1.import
- i.sentinel_1.mosaic
- i.sentinel_1.change

TODO: workflow for tests + delete README

### CODE REVIEW

File: i.sentinel_1.change/testsuite/test_i_sentinel1_change.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
+"""
+Name:       i.sentinel_1.change test
+Purpose:    Tests i.sentinel_1.change GRASS module.
+Author:     Guido Riembauer
+Copyright:  (C) 2021 Guido Riembauer, mundialis, and the GRASS
```
Comment (anikaweinmann) [inline i.sentinel_1.change/testsuite/test_i_sentinel1_change.py:5]:
```suggestion
Copyright:  (C) 2021-2022 mundialis GmbH & Co. KG, and the GRASS
```

**Review by mmacata — APPROVED** [review]

File: i.sentinel_1.change/i.sentinel_1.change.html [hunk 2]
```diff
@@ -0,0 +1,32 @@
+<h2>DESCRIPTION</h2>
+
+<em>i.sentinel_1.change</em> is a GRASS GIS addon that extracts areas of changed backscatter values from
+two Sentinel-1 input scenes. The input backscatter values are expected in dB scale.
+The addon calculates two polarisation ratios (date2_vv/date1_vv and date2_vh/date1_vh),
+smoothes them with a median filter in  <a href="r.neighbors.html">r.neighbors</a>
```
Comment (neteler) [inline i.sentinel_1.change/i.sentinel_1.change.html:6]:
```suggestion
smoothes them with a median filter in  <a href="https://grass.osgeo.org/grass-stable/manuals/r.neighbors.html">r.neighbors</a>
```

In addon manual pages, links to GRASS-core modules shall be full URLs.

File: i.sentinel_1.change/i.sentinel_1.change.html [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+i.sentinel_1.change date1_vv=Date1_Sigma0_VV_log date1_vh=Date1_Sigma0_VH_log date2_vv=Date2_Sigma0_VV_log date2_vh=Date2_Sigma0_VH_log output=changed_areas change_threshold=0.75 min_size=1.0 --o
+</pre></div>
+
+<h2>SEE ALSO</h2>
+<em>
+<a href="r.mapcalc.html">r.mapcalc</a>
+<a href="r.neighbors.html">r.neighbors</a>
+<a href="r.null.html">r.null</a>
+<a href="r.colors.html">r.colors</a>
+<a href="r.category.html">r.category</a>
```
Comment (neteler) [inline i.sentinel_1.change/i.sentinel_1.change.html:22]:
```suggestion
<a href="https://grass.osgeo.org/grass-stable/manuals/r.mapcalc.html">r.mapcalc</a>
<a href="https://grass.osgeo.org/grass-stable/manuals/r.neighbors.html">r.neighbors</a>
<a href="https://grass.osgeo.org/grass-stable/manuals/r.null.html">r.null</a>
<a href="https://grass.osgeo.org/grass-stable/manuals/r.colors.html">r.colors</a>
<a href="https://grass.osgeo.org/grass-stable/manuals/r.category.html">r.category</a>
```

File: i.sentinel_1.change/i.sentinel_1.change.py [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+#  (at your option) any later version.
+#
+#  This program is distributed in the hope that it will be useful,
+#  but WITHOUT ANY WARRANTY; without even the implied warranty of
+#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#  GNU General Public License for more details.
+#
+#############################################################################
+# %Module
+# % description: extracts changes from Sentinel-1 Scenes from two dates
```
Comment (neteler) [inline i.sentinel_1.change/i.sentinel_1.change.py:23]:
```suggestion
# % description: Extracts changes from Sentinel-1 Scenes from two dates.
```

File: i.sentinel_1.change/i.sentinel_1.change.py [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+# %option
+# % key: min_size
+# % type: double
+# % label: Minimum size of identified changed areas in ha
+# % answer: 1.0
+# %end
+
+# %option G_OPT_R_OUTPUT
+# % key: output
+# % label: Output raster containing areas of change
```
Comment (neteler) [inline i.sentinel_1.change/i.sentinel_1.change.py:67]:
```suggestion
# % label: Output raster map containing areas of change
```

File: i.sentinel_1.change/i.sentinel_1.change.py [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+# %end
+
+# %option G_OPT_R_INPUT
+# % key: date2_vv
+# % label: Input raster of second date in VV polarization (dB)
+# %end
+
+# %option G_OPT_R_INPUT
+# % key: date2_vh
+# % label: Input raster of second date in VH polarization (dB)
```
Comment (neteler) [inline i.sentinel_1.change/i.sentinel_1.change.py:47]:
```suggestion
# % label: Input raster map of first date in VV polarization (dB)
# %end

# %option G_OPT_R_INPUT
# % key: date1_vh
# % label: Input raster map of first date in VH polarization (dB)
# %end

# %option G_OPT_R_INPUT
# % key: date2_vv
# % label: Input raster map of second date in VV polarization (dB)
# %end

# %option G_OPT_R_INPUT
# % key: date2_vh
# % label: Input raster map of second date in VH polarization (dB)
```

File: i.sentinel_1.change/i.sentinel_1.change.py [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+    category_text = "\n".join(labels)
+
+    # assign labels
+    cat_proc = grass.feed_command("r.category", map=output, rules="-", separator=":")
+    cat_proc.stdin.write(category_text.encode())
+    cat_proc.stdin.close()
+    # feed_command does not wait until finished
+    cat_proc.wait()
+
+    grass.message(_(f"Successfully created output raster map {output}"))
```
Comment (neteler) [inline i.sentinel_1.change/i.sentinel_1.change.py:192]:
```suggestion
    grass.message(_(f"Successfully created output raster map <{output}>"))
```

Comment (neteler) [reply i.sentinel_1.change/i.sentinel_1.change.py:192]: Following:
https://trac.osgeo.org/grass/wiki/MessageStandardization

File: i.sentinel_1.download_asf/i.sentinel_1.download_asf.html [hunk 8]
```diff
@@ -0,0 +1,53 @@
+<h2>DESCRIPTION</h2>
+
+<em>i.sentinel_1.download_asf</em> searches and downloads SAR data from the Alaska
+Satellite Facility (ASF) using the <a href="https://asf.alaska.edu/api/">ASF-API</a>.
+The current GRASS GIS region is used to define the search area.
+Currently, <em>i.sentinel_1.download_asf</em> only allows the search and download of
+Sentinel-1 data. The benefit of this addon compared to
+<a href="i.sentinel.download">i.sentinel.download</a> is that the latter uses
```
Comment (neteler) [inline i.sentinel_1.download_asf/i.sentinel_1.download_asf.html:8]:
```suggestion
<a href="i.sentinel.download.html">i.sentinel.download</a> is that the latter uses
```

File: i.sentinel_1.html [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+
+The <em>i.sentinel_1</em> toolset consists of currently four modules:
+<dl>
+  <dt><a href="i.sentinel_1.download_asf.html">i.sentinel_1.download_asf</a></dt>
+  <dd>searches and downloads SAR data from the <a href="https://asf.alaska.edu/">Alaska Satellite Facility</a>
+  <dt><a href="i.sentinel_1.import">i.sentinel_1.import</a></dt>
+  <dd>preprocesses and imports Sentinel-1 GRD SAR data using ESA's <a href="https://step.esa.int/main/download/snap-download/">SNAP</a> software
+  <dt><a href="i.sentinel_1.mosaic">i.sentinel_1.mosaic</a></dt>
+  <dd>downloads and imports an entire Sentinel-1 GRD SAR coverage for the current region using <a href="i.sentinel_1.download_asf.html">i.sentinel_1.download_asf</a> and <a href="i.sentinel_1.import">i.sentinel_1.import</a>
+  <dt><a href="i.sentinel_1.change">i.sentinel_1.change</a></dt>
```
Comment (neteler) [inline i.sentinel_1.html:33]:
```suggestion
  <dt><a href="i.sentinel_1.import">i.sentinel_1.import</a></dt>
  <dd>preprocesses and imports Sentinel-1 GRD SAR data using ESA's <a href="https://step.esa.int/main/download/snap-download/">SNAP</a> software
  <dt><a href="i.sentinel_1.mosaic.html">i.sentinel_1.mosaic</a></dt>
  <dd>downloads and imports an entire Sentinel-1 GRD SAR coverage for the current region using <a href="i.sentinel_1.download_asf.html.html">i.sentinel_1.download_asf</a> and <a href="i.sentinel_1.import">i.sentinel_1.import</a>
  <dt><a href="i.sentinel_1.change.html">i.sentinel_1.change</a></dt>
```

File: i.sentinel_1.import/i.sentinel_1.import.html [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+and the SNAP-Python (snappy) interface needs to be configured (<a href="https://senbox.atlassian.net/wiki/spaces/SNAP/pages/50855941/Configure+Python+to+use+the+SNAP-Python+snappy+interface">instruction</a>)
+
+<h2>SEE ALSO</h2>
+<em>
+<a href="r.import.html">r.import</a>
+</em>
+
+<h2>AUTHORS</h2>
+
+Hajar Benelcadi, Guido Riembauer, mundialis
```
Comment (neteler) [inline i.sentinel_1.import/i.sentinel_1.import.html:39]:
```suggestion
Hajar Benelcadi, Guido Riembauer, <a href="https://www.mundialis.de/">mundialis</a>
```

File: i.sentinel_1.import/i.sentinel_1.import.py [hunk 11]
```diff
@@ -0,0 +1,352 @@
...
+#  This program is distributed in the hope that it will be useful,
+#  but WITHOUT ANY WARRANTY; without even the implied warranty of
+#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#  GNU General Public License for more details.
+#
+#############################################################################
+# %Module
+# % description: Uses snappy (SNAP ESA) to preprocess and import Sentinel 1 GRD data
+# % keyword: display
+# % keyword: raster
```
Comment (neteler) [inline i.sentinel_1.import/i.sentinel_1.import.py:28]:
```suggestion
# % keyword: raster
# % keyword: Sentinel
# % keyword: snappy
```

File: i.sentinel_1.import/i.sentinel_1.import.py [hunk 12]
```diff
@@ -0,0 +1,352 @@
...
+# % required: no
+# % multiple: no
+# % label: Path to locally stored DEM
+# % description: Path to locally stored DEM to perform terrain correction
+# %end
+
+# %flag
+# % key: s
+# % description: apply speckle filter
+# %end
```
Comment (neteler) [inline i.sentinel_1.import/i.sentinel_1.import.py:90]:
```suggestion
# % label: Path to the S-1 scene
# % description: Sentinel-1 data must be in .zip format
# %end

# %option
# % key: outpath
# % type: string
# % required: yes
# % multiple: no
# % label: Path to store BEAM-DIMAPs as intermediate products
# %end

# %option G_OPT_MEMORYMB
# %end

# %option
# % key: extent
# % type: string
# % required: no
# % multiple: no
# % label: Extent of imported raster map
# % options: input,region
# % answer: input
# %end

# %option
# % key: bandname
# % type: string
# % required: yes
# % multiple: yes
# %label: Bands to process
# % description: Bands to process
# % options: Sigma0_VV,Sigma0_VH,Gamma0_VV,Gamma0_VH
# % answer: Sigma0_VV,Sigma0_VH
# %end

# %option
# % key: external_dem
# % type: string
# % required: no
# % multiple: no
# % label: Path to locally stored DEM
# % description: Path to locally stored DEM to perform terrain correction
# %end

# %flag
# % key: s
# % description: Apply speckle filter
# %end
```

File: i.sentinel_1.import/i.sentinel_1.import.py [hunk 13]
```diff
@@ -0,0 +1,352 @@
...
+    # Hashmap is used to give us access to all JAVA oerators
+    GPF.getDefaultInstance().getOperatorSpiRegistry().loadOperatorSpis()
+    # HashMap = snappy.jpy.get_type("java.util.HashMap")
+    File = snappy.jpy.get_type("java.io.File")
+
+    if not os.path.exists(fileimage):
+        grass.fatal(_("File {name} not found.".format(name=fileimage)))
+
+    if not fileimage.endswith(".zip"):
+        grass.fatal(_("Input is not a S1 .zip file"))
```
Comment (neteler) [inline i.sentinel_1.import/i.sentinel_1.import.py:166]:
```suggestion
        grass.fatal(_("Input is not a Sentinel-1 .zip file"))
```

File: i.sentinel_1.mosaic/i.sentinel_1.mosaic.html [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+</ul>
+
+
+<h2>SEE ALSO</h2>
+
+<em>
+<a href="i.sentinel.download.html">i.sentinel.download</a>,
+<a href="i.sentinel_1.download_asf.html">i.sentinel_1.download_asf</a>,
+<a href="i.sentinel_1.import.html">i.sentinel_1.import</a>,
+<a href="r.patch.html">r.patch</a>
```
Comment (neteler) [inline i.sentinel_1.mosaic/i.sentinel_1.mosaic.html:58]:
```suggestion
<a href="https://grass.osgeo.org/grass-stable/manuals/r.patch.html">r.patch</a>
```

File: i.sentinel_1.mosaic/i.sentinel_1.mosaic.py [hunk 15]
```diff
@@ -0,0 +1,628 @@
...
+#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#  GNU General Public License for more details.
+#
+#############################################################################
+
+# %Module
+# % description: Searches for Sentinel-1 scenes based on current region, downloads, preprocesses, imports and patches.
+# % keyword: imagery
+# % keyword: satellite
+# % keyword: Sentinel
```
Comment (neteler) [inline i.sentinel_1.mosaic/i.sentinel_1.mosaic.py:30]:
```suggestion
# % keyword: Sentinel
# % keyword: SAR
```

File: i.sentinel_1.mosaic/i.sentinel_1.mosaic.py [hunk 16]
```diff
@@ -0,0 +1,628 @@
...
+# % label: Path to locally stored DEM
+# % description: Path to locally stored DEM to perform terrain correction
+# %end
+
+# %option G_OPT_MEMORYMB
+# %end
+
+# %flag
+# % key: s
+# % description: apply speckle filter
```
Comment (neteler) [inline i.sentinel_1.mosaic/i.sentinel_1.mosaic.py:102]:
```suggestion
# % description: Apply speckle filter
```

File: i.sentinel_1.import/i.sentinel_1.import.py [hunk 17]
```diff
@@ -0,0 +1,350 @@
...
+#############################################################################
+# %Module
+# % description: Uses snappy (SNAP ESA) to preprocess and import Sentinel 1 GRD data.
+# % keyword: display
+# % keyword: raster
+# % keyword: Sentinel 
+ # % keyword: SAR
+# % keyword: satellite
+# % keyword: snappy
+# %End
```
Comment (neteler) [inline i.sentinel_1.import/i.sentinel_1.import.py:33]:
```suggestion
# %Module
# % description: Uses snappy (SNAP ESA) to preprocess and import Sentinel 1 GRD data.
# % keyword: display
# % keyword: raster
# % keyword: Sentinel
# % keyword: SAR
# % keyword: satellite
# % keyword: snappy
# %End
```

File: i.sentinel_1.html [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+<body bgcolor="white">
+<div id="container">
+
+<a href="https://grass.osgeo.org/grass-stable/manuals/index.html"><img src="grass_logo.png" alt="GRASS logo"></a>
+<hr class="header">
+
+<h2>NAME</h2>
+
+<em><b>i.sentinel_1</b></em> - Toolset for downloading and processing of Copernicus
+Sentinel-1 data.
```
Comment (neteler) [inline i.sentinel_1.html:17]:
```suggestion
<em><b>i.sentinel_1</b></em> is a GRASS GIS addon toolset that searches, downloads, imports
and preprocesses SAR data from the <a href="https://asf.alaska.edu/">Alaska Satellite Facility</a>
using ESA's <a href="https://step.esa.int/main/download/snap-download/">SNAP</a>
software. If a region is covered by multiple Sentinel-1 GRD tiles, a mosaic can be created.
```

**Pull request merged**